src/msg/msg_types.h

```diff
@@ -175,21 +179,34 @@ WRITE_CLASS_ENCODER(ceph_sockaddr_storage)
 /*
  * encode sockaddr.ss_family as network byte order
  */
+static inline size_t get_sockaddr_length(int family) {
+    switch (family) {
+    case AF_INET:
+      return sizeof(struct sockaddr_in);
+    case AF_INET6:
+      return sizeof(struct sockaddr_in6);
+    }
+    return sizeof(struct sockaddr_storage);
+}
+
 static inline void encode(const sockaddr_storage& a, bufferlist& bl) {
 #if defined(__linux__)
   struct sockaddr_storage ss = a;
   ss.ss_family = htons(ss.ss_family);
   ::encode_raw(ss, bl);
 #elif defined(__FreeBSD__) || defined(__APPLE__)
+  // We do not copy ss_len into the buffer since the on-wire format 
+  // is compatible with the Linux sockaddr
   ceph_sockaddr_storage ss{};
   auto src = (unsigned char const *)&a;
   auto dst = (unsigned char *)&ss;
   src += sizeof(a.ss_len);
   ss.ss_family = a.ss_family;
   src += sizeof(a.ss_family);
   dst += sizeof(ss.ss_family);
-  const auto copy_size = std::min((unsigned char*)(&a + 1) - src,
-				  (unsigned char*)(&ss + 1) - dst);
+  // const auto copy_size = std::min((unsigned char*)(&a + 1) - src,
+  //				  (unsigned char*)(&ss + 1) - dst);
+  const auto copy_size = a.ss_len - sizeof(a.ss_family) - sizeof(a.ss_len);
```

Comment:
i think we should also memcpy the padding bytes to make sure all bits in `ss` are properly set. after this change, the behavior is broken.

Reply:
@tchaikov 
I guess we could, since in both platform the max size of the sockaddr (storage) is like 128 bytes.
On the otherhand we could init the block first with 0's??

But is the full sockaddr_storage send on the wire? What kind of information is there that we should be copying?

Reply:
for v1, yes. and per Sage, we are not sending the whole `sockaddr_storage ` over the wire. i think he replied you in the same PR.

Reply:
@tchaikov 
There is a wierd catch in this code, since Linux does not have sa_len, it does not suffer from that fact that a ANY_ADDRESS (aka all 0's) suffers from the fact that the length of the structure and sa_len are not equal. 
When sa_len exists it is 0 when the address is ANY_ADDRESS, so I'm splitting the size calculation of the structure from sa_len calculation, since they differ exactly on this topic.
Resulting in crashes from allocating `0xffffffff` sized buffers.